# 0036
# Cleaning Datasets

In [132]:
import pandas as pd 
pd.options.display.max_columns = 30

In [133]:
df = pd.read_csv("movies_metadata.csv", low_memory = False)


Explaination On low_memory = False :
By default, Pandas tries to determine the data type of each column in the CSV file as it reads the file. However, this can be a memory-intensive process, especially if the file is large or if there are many columns.

Setting low_memory=False tells Pandas to read the entire file into memory at once, rather than reading it in chunks. This can prevent memory errors that may occur when reading large files, but it can also be slower and more memory-intensive than reading in chunks.

If you're working with a large CSV file and you encounter memory errors or performance issues when reading the file, you may want to try setting low_memory=False. However, if you're working with a smaller file or if you have enough memory available, you may not need to set this argument.

It's important to note that setting low_memory=False can have performance implications, especially if you're working with a very large file or if you have limited memory available. In general, it's a good idea to test your code with and without this argument to see which approach works best for your specific use case.

In [134]:
df

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                                                  genres  \
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1      [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3      [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                         [{'id': 35, 'name': 'Comedy'}]   
...                                                  ...   
45461  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...   
45462                      [{'id': 18, 'name': 'Drama'}]   
45463  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   
45464                                                 []   
45465                                                 []   

                                   homepage      id    imdb_id  \
0      http://toystory.disney.com/toy-story     862  tt0114709   
1                                       NaN    8844  tt0113497   
2                                       NaN   15602  tt0113228   
3                                       NaN   31357  tt0114885   
4                                       NaN   11862  tt0113041   
...                                     ...     ...        ...   
45461  http://www.imdb.com/title/tt6209470/  439050  tt6209470   
45462                                   NaN  111109  tt2028550   
45463                                   NaN   67758  tt0303758   
45464                                   NaN  227506  tt0008536   
45465                                   NaN  461257  tt6980792   

      original_language               original_title  \
0                    en                    Toy Story   
1                    en                      Jumanji   
2                    en             Grumpier Old Men   
3                    en            Waiting to Exhale   
4                    en  Father of the Bride Part II   
...                 ...                          ...   
45461                fa                      رگ خواب   
45462                tl          Siglo ng Pagluluwal   
45463                en                     Betrayal   
45464                en          Satana likuyushchiy   
45465                en                     Queerama   

                                                overview popularity  \
0      Led by Woody, Andy's toys live happily in his ...  21.946943   
1      When siblings Judy and Peter discover an encha...  17.015539   
2      A family wedding reignites the ancient feud be...    11.7129   
3      Cheated on, mistreated and stepped on, the wom...   3.859495   
4      Just when George Banks has recovered from his ...   8.387519   
...                                                  ...        ...   
45461        Rising and falling between a man and woman.   0.072051   
45462  An artist struggles to finish his work while a...   0.178241   
45463  When one of her hits goes wrong, a professiona...   0.903007   
45464  In a small town live two brothers, one a minis...   0.003503   
45465  50 years after decriminalisation of homosexual...   0.163015   

                            poster_path  \

In [135]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [136]:
df['genres'][0]

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

In [137]:
df['belongs_to_collection'][0]

"{'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}"

Explaination on Output of previous cells:
The string is enclosed in square brackets, indicating that it represents a list. Within the square brackets, there are three dictionaries, each enclosed in curly braces. Each dictionary has two key-value pairs: 'id', which represents the unique identifier for the genre, and 'name', which represents the name of the genre.

In this case, the first dictionary has an 'id' of 16 and a 'name' of 'Animation'. The second dictionary has an 'id' of 35 and a 'name' of 'Comedy'. The third dictionary has an 'id' of 10751 and a 'name' of 'Family'.

This format is commonly used to represent nested or structured data in a single column of a Pandas DataFrame. To work with this data, you could use the ast.literal_eval() function to convert the string representation to a list of dictionarie

# 0037
#### Dropping Irrelevant Columns:
* dropping irrelevant columns in Pandas is a common task when working with data. Sometimes, a dataset may contain columns that are not useful for a particular analysis or that contain mostly missing values. In these cases, it can be useful to drop these columns to simplify the data and reduce memory usage. 

* To drop columns in Pandas, you can use the drop() method on a DataFrame. The drop() method takes one or more column names as arguments and returns a new DataFrame with those columns removed.

In [138]:
df.drop(columns = ['adult','imdb_id', 'original_title', 'video', 'homepage'], inplace = True)

In [139]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  4494 non-null   object 
 1   budget                 45466 non-null  object 
 2   genres                 45466 non-null  object 
 3   id                     45466 non-null  object 
 4   original_language      45455 non-null  object 
 5   overview               44512 non-null  object 
 6   popularity             45461 non-null  object 
 7   poster_path            45080 non-null  object 
 8   production_companies   45463 non-null  object 
 9   production_countries   45463 non-null  object 
 10  release_date           45379 non-null  object 
 11  revenue                45460 non-null  float64
 12  runtime                45203 non-null  float64
 13  spoken_languages       45460 non-null  object 
 14  status                 45379 non-null  object 
 15  ta

# 0038
# How To Handle Stringfied JSON Columns

#### Explaination of Stringfied JSON Columns:  

A stringified JSON column is a column in a dataset that contains JSON data in string format. JSON stands for JavaScript Object Notation, and it is a lightweight data interchange format that is easy for humans to read and write, and easy for machines to parse and generate.

In a stringified JSON column, the JSON data is represented as a string, rather than as a Python object or other data type. This can make it difficult to work with the JSON data, especially if you need to extract specific values or perform calculations on the data.

To work with stringified JSON columns in a Pandas dataset, you typically need to convert the stringified JSON to a Python object using the json module in Python


In [140]:
import json
import ast

In [141]:
json_colu = ['belongs_to_collection', 'genres', 'production_companies'
            , 'production_countries', 'spoken_languages']

In [142]:
df['belongs_to_collection'][0]

"{'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}"

In [143]:
json1 = '{"dogs": 3 , "cat":3}'
json1

'{"dogs": 3 , "cat":3}'

In [144]:
ast.literal_eval(json1)

{'dogs': 3, 'cat': 3}

We have a string called json1 that contains a JSON object in string format, and we want to convert this JSON object to a Python dictionary. One way to do this is to use the ast.literal_eval() function from the Python ast module.

In this code, we call ast.literal_eval() on json1, which safely evaluates the string as a Python expression and returns a dictionary with keys and values corresponding to the keys and values in the JSON object. In this case, the resulting dictionary will have a key 'dog' with a value of 3, and a key 'cat' with a value of 5.

#### Test on Dataset

In [145]:
df['genres'].apply(ast.literal_eval)[0]   

[{'id': 16, 'name': 'Animation'},
 {'id': 35, 'name': 'Comedy'},
 {'id': 10751, 'name': 'Family'}]

In [146]:
df_genres = df['genres'].apply(ast.literal_eval)
df_genres

0        [{'id': 16, 'name': 'Animation'}, {'id': 35, '...
1        [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...
2        [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...
3        [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...
4                           [{'id': 35, 'name': 'Comedy'}]
                               ...                        
45461    [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...
45462                        [{'id': 18, 'name': 'Drama'}]
45463    [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...
45464                                                   []
45465                                                   []
Name: genres, Length: 45466, dtype: object

# 0039
# How to Handle Stringfied JSON Columns P2

In [147]:
import numpy as np

#### Work on belongs_to_collection column

In [148]:
df['belongs_to_collection'][0]

"{'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}"

In [149]:
# check whether the Elements is Instance of str 
df['belongs_to_collection'].apply(lambda x :isinstance (x, str))

0         True
1        False
2         True
3        False
4         True
         ...  
45461    False
45462    False
45463    False
45464    False
45465    False
Name: belongs_to_collection, Length: 45466, dtype: bool

In [150]:
df['belongs_to_collection'] = df['belongs_to_collection'].apply(lambda x : ast.literal_eval(x) if isinstance (x, str) else np.nan)

#### Explaination of apply() & lambda :
The apply() method in Pandas can be used to apply a function to each element in a DataFrame or Series. In this case, we're applying a lambda function to the 'belongs_to_collection' column of the df DataFrame.

The lambda function checks if the input is a string using the isinstance() function. If the input is a string, it is assumed to be a JSON string, and the ast.literal_eval() function is used to convert the string to a Python object (in this case, a dictionary). If the input is not a string, the function returns np.nan.

The resulting output will be a new Series where each element is either a dictionary or np.nan, depending on the values in the 'belongs_to_collection' column of the df DataFrame. This can be useful for working with JSON data in a Pandas DataFrame, for example to extract specific values from a nested dictionary.

In [151]:
df['belongs_to_collection'][0]

{'id': 10194,
 'name': 'Toy Story Collection',
 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg',
 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}

In [152]:
# check the type of first record
type(df['belongs_to_collection'][0])

dict

#### Work on ```spoken_language``` column:

In [153]:
# check how its look like
df['spoken_languages'][0]

"[{'iso_639_1': 'en', 'name': 'English'}]"

In [154]:
# check how its look like
type(df['spoken_languages'][0])

str

In [155]:
df['spoken_languages'] = df['spoken_languages'].apply(lambda x : ast.literal_eval(x) if isinstance (x, str) else np.nan)

In [156]:
df['spoken_languages'][0]

[{'iso_639_1': 'en', 'name': 'English'}]

In [157]:
type(df['spoken_languages'][0])

list

#### Work on ```production_countries``` column:

In [158]:
df['production_companies'][0]

"[{'name': 'Pixar Animation Studios', 'id': 3}]"

In [159]:
type(df['production_countries'][0])

str

In [160]:
df['production_companies'] = df['production_companies'].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else np.nan)

In [161]:
df['production_companies'][0]

[{'name': 'Pixar Animation Studios', 'id': 3}]

In [162]:
type(df['production_companies'][0])

list

#### Work on ```genres``` column

In [163]:
df['genres'][0]

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

In [164]:
type(df['genres'][0])

str

In [165]:
df['genres'] = df['genres'].apply(lambda x : ast.literal_eval(x) if isinstance (x, str) else np.nan)

In [166]:
df['genres'][0]

[{'id': 16, 'name': 'Animation'},
 {'id': 35, 'name': 'Comedy'},
 {'id': 10751, 'name': 'Family'}]

In [167]:
type(df['genres'][0])

list

#### Check the DataFrame:

In [168]:
df.head()

belongs_to_collection    budget  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1                                                NaN  65000000   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3                                                NaN  16000000   
4  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres     id original_language  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...    862                en   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   8844                en   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  15602                en   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...  31357                en   
4                     [{'id': 35, 'name': 'Comedy'}]  11862                en   

                                            overview popularity  \
0  Led by Woody, Andy's toys live happily in his ...  21.946943   
1  When siblings Judy and Peter discover an encha...  17.015539   
2  A family wedding reignites the ancient feud be...    11.7129   
3  Cheated on, mistreated and stepped on, the wom...   3.859495   
4  Just when George Banks has recovered from his ...   8.387519   

                        poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg   
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0    101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0    127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0    106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  vote_average  vote_count  
0                    Toy Story           7.7      5415.0  
1                      Jumanji           6.9      2413.0  
2             Grumpier Old Men           6.5        92.0  
3            Waiting to Exhale           6.1        34.0  
4  Father of the Bride Part II           5.7       173.0

In [169]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  4494 non-null   object 
 1   budget                 45466 non-null  object 
 2   genres                 45466 non-null  object 
 3   id                     45466 non-null  object 
 4   original_language      45455 non-null  object 
 5   overview               44512 non-null  object 
 6   popularity             45461 non-null  object 
 7   poster_path            45080 non-null  object 
 8   production_companies   45463 non-null  object 
 9   production_countries   45463 non-null  object 
 10  release_date           45379 non-null  object 
 11  revenue                45460 non-null  float64
 12  runtime                45203 non-null  float64
 13  spoken_languages       45460 non-null  object 
 14  status                 45379 non-null  object 
 15  ta

# 0040:
# How to Flatten Nested Columns

In [170]:
df['belongs_to_collection'] = df['belongs_to_collection'].apply(lambda x: x['name'] if isinstance(x, dict) else np.nan)

The apply() method in Pandas can be used to apply a function to each element in a DataFrame or Series. In this case, we're applying a lambda function to the 'belongs_to_collection' column of the df DataFrame.

The lambda function checks if the input is a dictionary using the isinstance() function. If the input is a dictionary, it assumes that it represents a collection and returns the value associated with the 'name' key in the dictionary. If the input is not a dictionary, the function returns np.nan.

The resulting output will be a new Series where each element is either the value associated with the 'name' key in the original dictionary, or np.nan, depending on the values in the 'belongs_to_collection' column of the df DataFrame. This can be useful for extracting specific values from a nested dictionary in a Pandas DataFrame.

In [171]:
df['belongs_to_collection'].value_counts(dropna = False).head(20)

belongs_to_collection
NaN                                       40975
The Bowery Boys                              29
Totò Collection                              27
James Bond Collection                        26
Zatôichi: The Blind Swordsman                26
The Carry On Collection                      25
Pokémon Collection                           22
Charlie Chan (Sidney Toler) Collection       21
Godzilla (Showa) Collection                  16
Charlie Chan (Warner Oland) Collection       15
Dragon Ball Z (Movie) Collection             15
Uuno Turhapuro                               15
Monster High Collection                      14
The Land Before Time Collection              14
George Carlin Comedy Collection              13
Sharpe Collection                            13
Friday the 13th Collection                   12
The Amityville Horror Collection             12
Johan Falk GSI Collection                    12
Sherlock Holmes (1939 series)                12
Name: count, dtype

#### Flatten Data on ```Genres``` Column

In [172]:
df['genres'][0]

[{'id': 16, 'name': 'Animation'},
 {'id': 35, 'name': 'Comedy'},
 {'id': 10751, 'name': 'Family'}]

In [173]:
df['genres'] = df['genres'].apply(lambda x : "|".join(i['name'] for i in x))

The 'genres' column of the df DataFrame appears to contain a list of dictionaries, where each dictionary represents a genre with a 'name' key that contains the name of the genre. This format is commonly used to represent nested or structured data in a single column of a Pandas DataFrame.

The lambda function you're using with the apply() method is using a list comprehension to extract the value associated with the 'name' key from each dictionary in the list, and then joining those values into a single string separated by the '|' character. This creates a new column where each value is a string representing the genres associated with a particular entry in the DataFrame.

For example, if the 'genres' column of a particular row in the DataFrame contains the list [{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}], the lambda function will join the 'name' values together into the string 'Animation|Comedy|Family'.

The resulting output will be a new 'genres' column in the df DataFrame where each value is a string representing the genres associated with a particular entry in the DataFrame. This can be useful for filtering or grouping by genre in a Pandas DataFrame.

In [174]:
df['genres'][0]

'Animation|Comedy|Family'

In [175]:
df['genres'].value_counts(dropna = False).head(20)

genres
Drama                   5000
Comedy                  3621
Documentary             2723
                        2442
Drama|Romance           1301
Comedy|Drama            1135
Horror                   974
Comedy|Romance           930
Comedy|Drama|Romance     593
Drama|Comedy             532
Horror|Thriller          528
Drama|Thriller           497
Thriller                 465
Crime|Drama              430
Romance|Drama            343
Western                  318
Action|Thriller          301
Drama|Foreign            283
Action                   278
Drama|History            267
Name: count, dtype: int64

In [176]:
df.replace({"genres": ""}, np.nan, inplace=True)


as we can see in below output , we have 2442 rows without any genres, within previous cell , we did fill them with np.nan

Drama 5000
Comedy 3621
Documentary 2723
                        2442
Drama|Romance 1301
Comedy|Drama 1135
Horror 974
Comedy|Romance 930
Comedy|Drama|Romance 593
Drama|Comedy 532
Horror|Thriller 528
Drama|Thriller 497
Thriller 465
Crime|Drama 430
Romance|Drama 343
Western 318
Action|Thriller 301
Drama|Foreign 283
Action 278
Drama|History 267
Name: count, dtype: int64

In [177]:
df['genres'].value_counts(dropna = False).head(20)

genres
Drama                   5000
Comedy                  3621
Documentary             2723
NaN                     2442
Drama|Romance           1301
Comedy|Drama            1135
Horror                   974
Comedy|Romance           930
Comedy|Drama|Romance     593
Drama|Comedy             532
Horror|Thriller          528
Drama|Thriller           497
Thriller                 465
Crime|Drama              430
Romance|Drama            343
Western                  318
Action|Thriller          301
Drama|Foreign            283
Action                   278
Drama|History            267
Name: count, dtype: int64

#### Flatten Data on ```spoken_language```

In [178]:
df['spoken_languages'][0]

[{'iso_639_1': 'en', 'name': 'English'}]

In [179]:
df['spoken_languages'] = df['spoken_languages'].apply(lambda x : "|".join(i['name'] for i in x) if isinstance (x, list) else np.nan)

In [180]:
df['spoken_languages'].value_counts(dropna = False).head(20)

spoken_languages
English             22395
                     3952
Français             1853
日本語                  1289
Italiano             1218
Español               902
Pусский               807
Deutsch               762
English|Français      681
English|Español       572
हिन्दी                481
English|Deutsch       462
한국어/조선말               425
普通话                   347
English|Italiano      326
svenska               311
No Language           303
Português             275
suomi                 275
Polski                213
Name: count, dtype: int64

In [181]:
df.replace({"spoken_languages": ""}, np.nan, inplace=True)

In [182]:
df['spoken_languages'].value_counts(dropna = False).head(20)

spoken_languages
English             22395
NaN                  3958
Français             1853
日本語                  1289
Italiano             1218
Español               902
Pусский               807
Deutsch               762
English|Français      681
English|Español       572
हिन्दी                481
English|Deutsch       462
한국어/조선말               425
普通话                   347
English|Italiano      326
svenska               311
No Language           303
Português             275
suomi                 275
Polski                213
Name: count, dtype: int64

#### Flatten Data on ```production_countries``` column

In [183]:
df['production_countries'][0]

"[{'iso_3166_1': 'US', 'name': 'United States of America'}]"

In [184]:
df['production_countries'] = df['production_countries'].apply(lambda x : "|".join(i['name'] for i in x) if isinstance(x, list) else np.nan)

In [185]:
#df['production_countries'] = df['production_countries'].replace("", np.nan)


In [186]:
df['production_countries'].value_counts(dropna = False).head(20)

production_countries
NaN    45466
Name: count, dtype: int64

#### Flatten Data on ```production_companies``` column

In [187]:
df['production_companies'][5]

[{'name': 'Regency Enterprises', 'id': 508},
 {'name': 'Forward Pass', 'id': 675},
 {'name': 'Warner Bros.', 'id': 6194}]

In [188]:
df['production_companies'] = df['production_companies'].apply(lambda x: "|".join(i['name'] for i in x) if isinstance(x, list) else np.nan)

In [189]:
df['production_companies'].value_counts(dropna = False).head(20)

production_companies
                                          11875
Metro-Goldwyn-Mayer (MGM)                   742
Warner Bros.                                540
Paramount Pictures                          505
Twentieth Century Fox Film Corporation      439
Universal Pictures                          320
RKO Radio Pictures                          247
Columbia Pictures Corporation               207
Columbia Pictures                           146
Mosfilm                                     145
Walt Disney Pictures                         85
Universal International Pictures (UI)        82
New Line Cinema                              75
Walt Disney Productions                      75
Shaw Brothers                                71
Touchstone Pictures                          70
Toho Company                                 65
TriStar Pictures                             62
Orion Pictures                               61
Hammer Film Productions                      60
Name: count, dtype:

In [190]:
df.replace("production_companies", np.nan, inplace = True)
df['production_companies'].value_counts(dropna = False).head(20)

production_companies
                                          11875
Metro-Goldwyn-Mayer (MGM)                   742
Warner Bros.                                540
Paramount Pictures                          505
Twentieth Century Fox Film Corporation      439
Universal Pictures                          320
RKO Radio Pictures                          247
Columbia Pictures Corporation               207
Columbia Pictures                           146
Mosfilm                                     145
Walt Disney Pictures                         85
Universal International Pictures (UI)        82
New Line Cinema                              75
Walt Disney Productions                      75
Shaw Brothers                                71
Touchstone Pictures                          70
Toho Company                                 65
TriStar Pictures                             62
Orion Pictures                               61
Hammer Film Productions                      60
Name: count, dtype:

In [191]:
df.isna().sum()

belongs_to_collection    40975
budget                       0
genres                    2442
id                           0
original_language           11
overview                   954
popularity                   5
poster_path                386
production_companies         6
production_countries     45466
release_date                87
revenue                      6
runtime                    263
spoken_languages          3958
status                      87
tagline                  25054
title                        6
vote_average                 6
vote_count                   6
dtype: int64

In [192]:
pd.read_csv("movies_metadata.csv", low_memory=False).isna().sum()

adult                        0
belongs_to_collection    40972
budget                       0
genres                       0
homepage                 37684
id                           0
imdb_id                     17
original_language           11
original_title               0
overview                   954
popularity                   5
poster_path                386
production_companies         3
production_countries         3
release_date                87
revenue                      6
runtime                    263
spoken_languages             6
status                      87
tagline                  25054
title                        6
video                        6
vote_average                 6
vote_count                   6
dtype: int64

# 0041
# Cleaning Numberical Columns
***Budget & BoxOffice Col***

In [193]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  4491 non-null   object 
 1   budget                 45466 non-null  object 
 2   genres                 43024 non-null  object 
 3   id                     45466 non-null  object 
 4   original_language      45455 non-null  object 
 5   overview               44512 non-null  object 
 6   popularity             45461 non-null  object 
 7   poster_path            45080 non-null  object 
 8   production_companies   45460 non-null  object 
 9   production_countries   0 non-null      float64
 10  release_date           45379 non-null  object 
 11  revenue                45460 non-null  float64
 12  runtime                45203 non-null  float64
 13  spoken_languages       41508 non-null  object 
 14  status                 45379 non-null  object 
 15  ta

#### Explaination of `pd.to_numeric`()
The pd.to_numeric() function in Pandas can be used to convert a Series or DataFrame column to a numeric data type. In this case, you're passing the 'budget' column of the df DataFrame to the function, along with the argument errors='coerce'. The errors argument tells the function how to handle any non-numeric values in the column.

When errors='coerce', any non-numeric values in the column will be converted to NaN, which stands for "Not a Number". This is useful for handling missing or invalid data in a column without raising an error.

The resulting output will be a new 'budget' column in the df DataFrame where each value is a numeric data type, or NaN if the original value could not be converted to a numeric data type. This can be useful for performing mathematical operations or comparing values in a Pandas DataFrame.

In [194]:
df['budget'] = pd.to_numeric(df['budget'], errors='coerce')

#### **Replace the 0 with np.nan**

budget  
`0.0           36573  `
5000000.0       286  
10000000.0      259  
20000000.0      243  
2000000.0       242  

In [195]:
df['budget'] = df['budget'].replace(0, np.nan)

In [196]:
df['budget'].value_counts(dropna = False)

budget
NaN            36576
5000000.0        286
10000000.0       259
20000000.0       243
2000000.0        242
               ...  
152000000.0        1
1500.0             1
82000.0            1
1282545.0          1
2153912.0          1
Name: count, Length: 1223, dtype: int64

#### Conver values of `Budget` to Smaller Number by Dividing 1000000


In [197]:
df['budget'] = df['budget'].div(1000000)

In [198]:
df['budget'].value_counts(dropna = False)

budget
NaN           36576
5.000000        286
10.000000       259
20.000000       243
2.000000        242
              ...  
152.000000        1
0.001500          1
0.082000          1
1.282545          1
2.153912          1
Name: count, Length: 1223, dtype: int64

#### `Revenue` Column

In [199]:
df['revenue'].value_counts(dropna=False)

revenue
0.0           38052
12000000.0       20
11000000.0       19
10000000.0       19
2000000.0        18
              ...  
320395.0          1
50656.0           1
1328612.0         1
1268793.0         1
1413000.0         1
Name: count, Length: 6864, dtype: int64

In [200]:
df['revenue'] = df['revenue'].replace(0, np.nan)
df['revenue'] = df['revenue'].div(1000000)

alling the div() method on the 'revenue' column of the df DataFrame, and passing the argument 1000000. This divides each value in the 'revenue' column by 1,000,000, effectively converting the revenue values from their original units to millions of dollars.

#### Changing the Name of `Revenue` & `Budget` Column

In [201]:
df.rename(columns={'revenue' : 'revenue_musd', 'budget' : 'budget_musd'}, inplace = True)

In [202]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  4491 non-null   object 
 1   budget_musd            8890 non-null   float64
 2   genres                 43024 non-null  object 
 3   id                     45466 non-null  object 
 4   original_language      45455 non-null  object 
 5   overview               44512 non-null  object 
 6   popularity             45461 non-null  object 
 7   poster_path            45080 non-null  object 
 8   production_companies   45460 non-null  object 
 9   production_countries   0 non-null      float64
 10  release_date           45379 non-null  object 
 11  revenue_musd           7408 non-null   float64
 12  runtime                45203 non-null  float64
 13  spoken_languages       41508 non-null  object 
 14  status                 45379 non-null  object 
 15  ta

#### `df.rename` function :
***The rename() method in Pandas can be used to rename one or more columns (or row labels) in a DataFrame. In this case, you're calling the rename() method on the df DataFrame, and passing the argument columns={'revenue': 'revenue_musd', 'budget': 'budget_musd'} to specify the new names for the 'revenue' and 'budget' columns.***

***The columns argument is a dictionary where the keys are the original column names, and the values are the new column names. In this case, the original 'revenue' column is being renamed to 'revenue_musd', and the original 'budget' column is being renamed to 'budget_musd'.***

# 0042
# Cleaning Numerical Columns P2

In [203]:
df['runtime'].value_counts(dropna = False).head(20)
                           

runtime
90.0     2556
0.0      1558
100.0    1470
95.0     1412
93.0     1214
96.0     1104
92.0     1080
94.0     1062
91.0     1057
88.0     1032
97.0     1027
85.0     1024
98.0     1019
105.0    1002
89.0      958
87.0      919
110.0     850
86.0      846
99.0      794
102.0     791
Name: count, dtype: int64

One of the most occured values of runtime is `0` which is not reallistic in this df. it is better to change 0 values to `np.nan`  

runtime
90.0     2556    
`0.0      1558`  
100.0    1470  
95.0     1412  
93.0     1214  
96.0     1104  
92.0     1080  

In [204]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  4491 non-null   object 
 1   budget_musd            8890 non-null   float64
 2   genres                 43024 non-null  object 
 3   id                     45466 non-null  object 
 4   original_language      45455 non-null  object 
 5   overview               44512 non-null  object 
 6   popularity             45461 non-null  object 
 7   poster_path            45080 non-null  object 
 8   production_companies   45460 non-null  object 
 9   production_countries   0 non-null      float64
 10  release_date           45379 non-null  object 
 11  revenue_musd           7408 non-null   float64
 12  runtime                45203 non-null  float64
 13  spoken_languages       41508 non-null  object 
 14  status                 45379 non-null  object 
 15  ta

#### Covert `id` Column Dtype to Numeric

as we can see in output of `df.info` the data type related to column `id` is object which is not proper dtype.  
in this case we will change it to number using `pd.to_numeric()`

In [205]:
df['id'] = pd.to_numeric(df['id'], errors = 'coerce')

In [206]:
df['id'].value_counts(dropna = False).head(20)

id
141971.0    3
NaN         3
69234.0     2
13209.0     2
109962.0    2
132641.0    2
105045.0    2
15028.0     2
14788.0     2
4912.0      2
168538.0    2
97995.0     2
22649.0     2
12600.0     2
10991.0     2
5511.0      2
159849.0    2
152795.0    2
11115.0     2
99080.0     2
Name: count, dtype: int64

***WE Have Dupelicate And Missing Values in `id` Column***

In [207]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  4491 non-null   object 
 1   budget_musd            8890 non-null   float64
 2   genres                 43024 non-null  object 
 3   id                     45463 non-null  float64
 4   original_language      45455 non-null  object 
 5   overview               44512 non-null  object 
 6   popularity             45461 non-null  object 
 7   poster_path            45080 non-null  object 
 8   production_companies   45460 non-null  object 
 9   production_countries   0 non-null      float64
 10  release_date           45379 non-null  object 
 11  revenue_musd           7408 non-null   float64
 12  runtime                45203 non-null  float64
 13  spoken_languages       41508 non-null  object 
 14  status                 45379 non-null  object 
 15  ta

as we can see in output of `df.info` the data type related to column `popularity` is object which is not proper dtype.  
in this case we will change it to number using `pd.to_numeric()`

In [208]:
df['popularity'] = pd.to_numeric( df['popularity'], errors = 'coerce')

In [209]:
df['popularity'].value_counts(dropna = False).head(20)

popularity
0.000000    66
0.000001    56
0.000308    43
0.000220    40
0.000844    38
0.001177    38
0.000578    38
0.002001    28
0.003013    21
0.003530    19
0.001393    19
0.002353    18
0.036471    18
0.000603    16
0.001586    15
0.004425    14
0.000431    13
0.001021    13
0.004706    12
0.001247    11
Name: count, dtype: int64

### Analysis on `vote_counts` and `vote_average` Columns

In [210]:
df['vote_count'].value_counts(dropna = False).head(20)

vote_count
1.0     3264
2.0     3132
0.0     2899
3.0     2787
4.0     2480
5.0     2097
6.0     1747
7.0     1570
8.0     1359
9.0     1194
10.0    1171
11.0     944
12.0     859
13.0     733
14.0     700
15.0     674
16.0     601
17.0     554
18.0     497
20.0     463
Name: count, dtype: int64

In [211]:
df['vote_average'].value_counts(dropna = False).head(20)

vote_average
0.0    2998
6.0    2468
5.0    2001
7.0    1886
6.5    1722
6.3    1603
5.5    1381
5.8    1369
6.4    1350
6.7    1342
6.8    1324
6.1    1281
6.6    1263
6.2    1253
5.9    1196
5.3    1082
5.7    1046
6.9    1037
5.6    1006
7.3    1000
Name: count, dtype: int64

In [212]:
df.loc[df['vote_count'] == 0 , 'vote_count']

83       0.0
107      0.0
126      0.0
132      0.0
137      0.0
        ... 
45432    0.0
45434    0.0
45452    0.0
45464    0.0
45465    0.0
Name: vote_count, Length: 2899, dtype: float64

The loc[] method in Pandas can be used to select specific rows and columns in a DataFrame based on their labels. In this case, you're using boolean indexing to select the rows in the 'vote_average' column of the df DataFrame where the value is 0.

You're then accessing the 'vote_average' column of the selected rows using the syntax ['vote_average']. This returns a Series containing only the values in the 'vote_average' column of the selected rows.

The resulting output will be a new Series where the index labels are the same as the original DataFrame df, but containing only the 'vote_average' values from the rows where the 'vote_average' value is 0.

# 0043
# How to Clean Columns With Datetime Information
***`release_date` Column***

In [213]:
df['release_date'][0]

'1995-10-30'

In [214]:
df['release_date'] = pd.to_datetime(df['release_date'],errors='coerce')

In [215]:
df['release_date'].value_counts(dropna = False).head(20)

release_date
2008-01-01    136
2009-01-01    121
2007-01-01    118
2005-01-01    111
2006-01-01    101
2002-01-01     96
2004-01-01     90
NaT            90
2001-01-01     84
2003-01-01     76
1997-01-01     69
2010-01-01     68
1999-01-01     67
1998-01-01     67
2000-01-01     64
1988-01-01     63
1987-01-01     63
1994-01-01     61
1989-01-01     60
2011-01-01     56
Name: count, dtype: int64

df.info()

***It is shown that `release_date` column's dtype has been changed to datetime by using `pd.to_datatime()` function***

# 004
# How to Clean String Text Columns
***`original_language` and `overview`***

In [216]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   belongs_to_collection  4491 non-null   object        
 1   budget_musd            8890 non-null   float64       
 2   genres                 43024 non-null  object        
 3   id                     45463 non-null  float64       
 4   original_language      45455 non-null  object        
 5   overview               44512 non-null  object        
 6   popularity             45460 non-null  float64       
 7   poster_path            45080 non-null  object        
 8   production_companies   45460 non-null  object        
 9   production_countries   0 non-null      float64       
 10  release_date           45376 non-null  datetime64[ns]
 11  revenue_musd           7408 non-null   float64       
 12  runtime                45203 non-null  float64       
 13  s

In [217]:
df['original_language'].value_counts(dropna = False).head(20)

original_language
en    32269
fr     2438
it     1529
ja     1350
de     1080
es      994
ru      826
hi      508
ko      444
zh      409
sv      384
pt      316
cn      313
fi      297
nl      248
da      225
pl      219
tr      150
cs      130
el      113
Name: count, dtype: int64

In [218]:
df['title']

0                          Toy Story
1                            Jumanji
2                   Grumpier Old Men
3                  Waiting to Exhale
4        Father of the Bride Part II
                    ...             
45461                         Subdue
45462            Century of Birthing
45463                       Betrayal
45464               Satan Triumphant
45465                       Queerama
Name: title, Length: 45466, dtype: object

In [219]:
df['title'].value_counts(dropna = False).head(10)

title
Cinderella              11
Alice in Wonderland      9
Hamlet                   9
Beauty and the Beast     8
Les Misérables           8
The Three Musketeers     7
A Christmas Carol        7
Blackout                 7
Treasure Island          7
Jane Eyre                6
Name: count, dtype: int64

#### ***Cleaning Values on Overview Column***

In [220]:
df['overview'][0]

"Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy's heart, Woody plots against Buzz. But when circumstances separate Buzz and Woody from their owner, the duo eventually learns to put aside their differences."

In [221]:
df['overview'].value_counts(dropna = False).head(20)

overview
NaN                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              954
No overview found.                                                                                                                                        

In [222]:
df['overview'][0]

"Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy's heart, Woody plots against Buzz. But when circumstances separate Buzz and Woody from their owner, the duo eventually learns to put aside their differences."

In [223]:
df.replace({'overview': {'No overview found.': np.nan}}, inplace=True)


In [224]:
df['overview'].value_counts(dropna = False).head(20)

overview
NaN                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              1087
No Overview                                                                                                                                              

In [225]:
inplace =True

In [226]:
df.replace({'overview':{'No Overview', np.nan}}, inplace = True)

In [227]:
df.replace({'overview':{'No movie overview available.', np.nan}}, inplace = True)

In [228]:
df.replace({'overview':{" ", np.nan}}, inplace = True)

In [229]:
df.replace({'overview':{'No overview yet', np.nan}}, inplace = True)

In [230]:
df['overview'].value_counts(dropna = False).head(20)

overview
NaN                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              1087
No Overview                                                                                                                                              

#### Cleaning `tagline` column

In [231]:
df['tagline'].value_counts(dropna = False).head(20)

tagline
NaN                                                           25054
Based on a true story.                                            7
-                                                                 4
Be careful what you wish for.                                     4
Trust no one.                                                     4
Some doors should never be opened.                                3
Drama                                                             3
Know Your Enemy                                                   3
Which one is the first to return - memory or the murderer?        3
There are two sides to every love story.                          3
The end is near.                                                  3
A Love Story                                                      3
There is no turning back                                          3
Who is John Galt?                                                 3
How far would you go?                   

In [232]:
df.replace({'tagline':{'-', np.nan}}, inplace = True)

In [233]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   belongs_to_collection  4491 non-null   object        
 1   budget_musd            8890 non-null   float64       
 2   genres                 43024 non-null  object        
 3   id                     45463 non-null  float64       
 4   original_language      45455 non-null  object        
 5   overview               44379 non-null  object        
 6   popularity             45460 non-null  float64       
 7   poster_path            45080 non-null  object        
 8   production_companies   45460 non-null  object        
 9   production_countries   0 non-null      float64       
 10  release_date           45376 non-null  datetime64[ns]
 11  revenue_musd           7408 non-null   float64       
 12  runtime                45203 non-null  float64       
 13  s

# 0045
# Removing Duplicate Rows

In [234]:
df[df.duplicated( keep = False)].sort_values( by ='id')

belongs_to_collection  budget_musd  \
7345                    NaN          NaN   
9165                    NaN          NaN   
14012                   NaN          NaN   
24844                   NaN          NaN   
22151                   NaN          NaN   
14000                   NaN          NaN   
9327                    NaN          NaN   
8068                    NaN          NaN   
17229                   NaN          NaN   
23044                   NaN          NaN   
15702                   NaN          NaN   
29374                   NaN          NaN   
40276                   NaN          NaN   
13946                   NaN          NaN   
16167                   NaN          NaN   
38871                   NaN          NaN   
676                     NaN          NaN   
1465                    NaN          NaN   
21165                   NaN          NaN   
19890                   NaN          NaN   
13261                   NaN          NaN   
16764                   NaN          NaN   
13375                   NaN          NaN   
19925                   NaN          NaN   
21854                   NaN          NaN   
13220          Why We Fight          NaN   
35798          Why We Fight          NaN   
13603                   NaN          NaN   
28860                   NaN          NaN   
24164                   NaN          NaN   
45265                   NaN          NaN   
33184                   NaN         0.98   
40040                   NaN         0.98   

                                         genres        id original_language  \
7345                       Crime|Drama|Thriller    5511.0                fr   
9165                       Crime|Drama|Thriller    5511.0                fr   
14012                              Comedy|Drama   11115.0                en   
24844                              Comedy|Drama   11115.0                en   
22151             Action|Horror|Science Fiction   18440.0                en   
14000             Action|Horror|Science Fiction   18440.0                en   
9327   Adventure|Animation|Drama|Action|Foreign   23305.0                en   
8068   Adventure|Animation|Drama|Action|Foreign   23305.0                en   
17229                                     Drama   25541.0                da   
23044                                     Drama   25541.0                da   
15702                             Drama|Foreign   42495.0                en   
29374                             Drama|Foreign   42495.0                en   
40276                                    Comedy   97995.0                en   
13946                                    Comedy   97995.0                en   
16167            Action|Drama|Romance|Adventure   99080.0                en   
38871            Action|Drama|Romance|Adventure   99080.0                en   
676                               Drama|Romance  105045.0                de   
1465                              Drama|Romance  105045.0                de   
21165             Fantasy|Drama|Science Fiction  119916.0                en   
19890             Fantasy|Drama|Science Fiction  119916.0                en   
13261                          Thriller|Mystery  141971.0                fi   
16764                          Thriller|Mystery  141971.0                fi   
13375                          Thriller|Mystery  141971.0                fi   
19925           Drama|Science Fiction|Animation  152795.0                en   
21854           Drama|Science Fiction|Animation  152795.0                en   
13220                               Documentary  159849.0                en   
35798                               Documentary  159849.0                en   
13603                              Drama|Comedy  168538.0                en   
28860                              Drama|Comedy  168538.0                en   
24164                              Comedy|Drama  265189.0                sv   
45265                              Comedy|Drama  265189.0   

In [235]:
df.drop_duplicates(inplace = True)

In [236]:
df[df.duplicated(subset = 'id', keep=False)].sort_values(by = 'id')

belongs_to_collection  budget_musd  \
5865                    NaN    30.000000   
33826                   NaN    30.000000   
44821    Pokémon Collection    16.000000   
4114     Pokémon Collection    16.000000   
5535     Pokémon Collection          NaN   
44826    Pokémon Collection          NaN   
15765                   NaN     0.002500   
11342                   NaN     0.002500   
12066                   NaN     1.600000   
10419                   NaN     1.600000   
33743                   NaN    26.000000   
5130                    NaN    26.000000   
949                     NaN     0.000004   
15074                   NaN     0.000004   
9576                    NaN    10.000000   
26625                   NaN    10.000000   
11155                   NaN    40.000000   
20843                   NaN    40.000000   
21116                   NaN          NaN   
2564                    NaN          NaN   
5710                    NaN          NaN   
20899                   NaN          NaN   
23534                   NaN     3.512454   
4356                    NaN     3.512454   
838                     NaN          NaN   
30001                   NaN          NaN   
19730                   NaN          NaN   
29503                   NaN          NaN   
35587                   NaN          NaN   

                                                  genres        id  \
5865                 Comedy|Crime|Drama|Romance|Thriller    4912.0   
33826                Comedy|Crime|Drama|Romance|Thriller    4912.0   
44821          Adventure|Fantasy|Animation|Action|Family   10991.0   
4114           Adventure|Fantasy|Animation|Action|Family   10991.0   
5535   Adventure|Fantasy|Animation|Science Fiction|Fa...   12600.0   
44826  Adventure|Fantasy|Animation|Science Fiction|Fa...   12600.0   
15765                               Drama|Comedy|Foreign   13209.0   
11342                               Drama|Comedy|Foreign   13209.0   
12066                                Drama|Crime|Mystery   14788.0   
10419                                Drama|Crime|Mystery   14788.0   
33743          Adventure|Family|Science Fiction|Thriller   15028.0   
5130           Adventure|Family|Science Fiction|Thriller   15028.0   
949                                    Drama|Romance|War   22649.0   
15074                                  Drama|Romance|War   22649.0   
9576                 Drama|Horror|Music|Romance|TV Movie   69234.0   
26625                Drama|Horror|Music|Romance|TV Movie   69234.0   
11155                                    Adventure|Drama   77221.0   
20843                                    Adventure|Drama   77221.0   
21116                                        Documentary   84198.0   
2564                                         Documentary   84198.0   
5710                                               Drama  109962.0   
20899                                              Drama  109962.0   
23534                                              Drama  110428.0   
4356                                               Drama  110428.0   
838                                                Drama  132641.0   
30001                                              Drama  132641.0   
19730  Carousel Productions|Vision View Entertainment...       NaN   
29503  Aniplex|GoHands|BROSTA TV|Mardock Scramble Pro...       NaN   
35587  Odyssey Media|Pulser Productions|Rogue State|T...       NaN   

      original_language                                           overview  \
5865                 en  Television made him famous, but his biggest hi...   
33826                en  Television made him famous, but his biggest hi...   
44821                ja  When Molly Hale's sadness of her father's disa...   
4114                 ja  When Molly Hale's sadness of her father's disa...   
5535                 ja  All your favorite Pokémon characters are back,...   
44826                ja  All your favorite Pokémon characters are back,...   
15765                fa  Since women a

#### Explaination On `df.duplicated()` : 
The duplicated() method in Pandas can be used to identify duplicate rows in a DataFrame based on one or more columns. In this case, you're calling the duplicated() method on the df DataFrame, and passing the arguments subset='id' to specify that you want to consider only the 'id' column when identifying duplicates, and keep=False to keep all instances of duplicated rows in the DataFrame, rather than just the first or last instance.

The resulting output will be a Boolean Series where each value is True if the corresponding row is a duplicate based on the 'id' column, and False otherwise.

In [128]:
df.drop_duplicates(subset = 'id', inplace = True)

In [129]:
df['id'].value_counts(dropna = False)

id
461257.0    1
862.0       1
8844.0      1
15602.0     1
31357.0     1
           ..
1408.0      1
10858.0     1
21032.0     1
12110.0     1
9087.0      1
Name: count, Length: 45434, dtype: int64

# Final Steps of Cleaning Df

In [238]:
df['status'].value_counts()

status
Released           44998
Rumored              229
Post Production       98
In Production         20
Planned               15
Canceled               2
Name: count, dtype: int64

In [239]:
df = df.loc[df['status'] == 'Released'].copy()

The loc[] method in Pandas can be used to select specific rows and columns in a DataFrame based on their labels. In this case, you're using boolean indexing to select the rows in the 'status' column of the df DataFrame where the value is 'Released'.

You're then calling the copy() method on the resulting DataFrame to create a copy of the selected rows. This is useful because modifying a view of a DataFrame (i.e., a slice of the original DataFrame) can sometimes modify the original DataFrame as well, which can lead to unexpected behavior.

The resulting output will be a new DataFrame containing only the rows where the 'status' value is 'Released'. This can be useful for filtering out any rows that are not relevant to your analysis.

In [240]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 44998 entries, 0 to 45465
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   belongs_to_collection  4465 non-null   object        
 1   budget_musd            8864 non-null   float64       
 2   genres                 42614 non-null  object        
 3   id                     44998 non-null  float64       
 4   original_language      44988 non-null  object        
 5   overview               43945 non-null  object        
 6   popularity             44998 non-null  float64       
 7   poster_path            44625 non-null  object        
 8   production_companies   44998 non-null  object        
 9   production_countries   0 non-null      float64       
 10  release_date           44920 non-null  datetime64[ns]
 11  revenue_musd           7392 non-null   float64       
 12  runtime                44747 non-null  float64       
 13  spoken

In [241]:
df.drop(columns=['status'],inplace=True)

In [242]:
col = ['id', 'title', 'tagline', 'release_date', 'genres', 'belongs_to_collection',
      'original_language', 'budget_musd', 'revenue_musd', 'production_companies', 
      'production_countries', 'vote_count', 'vote_average', 'popularity', 'runtime', 'overview',
      'spoken_languages', 'poster_path']

In [243]:
df=df.loc[:,col]

In [244]:
df.head(5)

id                        title  \
0    862.0                    Toy Story   
1   8844.0                      Jumanji   
2  15602.0             Grumpier Old Men   
3  31357.0            Waiting to Exhale   
4  11862.0  Father of the Bride Part II   

                                             tagline release_date  \
0                                                NaN   1995-10-30   
1          Roll the dice and unleash the excitement!   1995-12-15   
2  Still Yelling. Still Fighting. Still Ready for...   1995-12-22   
3  Friends are the people who let you be yourself...   1995-12-22   
4  Just When His World Is Back To Normal... He's ...   1995-02-10   

                     genres           belongs_to_collection original_language  \
0   Animation|Comedy|Family            Toy Story Collection                en   
1  Adventure|Fantasy|Family                             NaN                en   
2            Romance|Comedy       Grumpy Old Men Collection                en   
3      Comedy|Drama|Romance                             NaN                en   
4                    Comedy  Father of the Bride Collection                en   

   budget_musd  revenue_musd  \
0         30.0    373.554033   
1         65.0    262.797249   
2          NaN           NaN   
3         16.0     81.452156   
4          NaN     76.578911   

                                production_companies  production_countries  \
0                            Pixar Animation Studios                   NaN   
1  TriStar Pictures|Teitler Film|Interscope Commu...                   NaN   
2                        Warner Bros.|Lancaster Gate                   NaN   
3             Twentieth Century Fox Film Corporation                   NaN   
4          Sandollar Productions|Touchstone Pictures                   NaN   

   vote_count  vote_average  popularity  runtime  \
0      5415.0           7.7   21.946943     81.0   
1      2413.0           6.9   17.015539    104.0   
2        92.0           6.5   11.712900    101.0   
3        34.0           6.1    3.859495    127.0   
4       173.0           5.7    8.387519    106.0   

                                            overview  spoken_languages  \
0  Led by Woody, Andy's toys live happily in his ...           English   
1  When siblings Judy and Peter discover an encha...  English|Français   
2  A family wedding reignites the ancient feud be...           English   
3  Cheated on, mistreated and stepped on, the wom...           English   
4  Just when George Banks has recovered from his ...           English   

                        poster_path  
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg  
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg  
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg  
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg  
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg

In [246]:
df.reset_index(drop = True,inplace = True)

Explaination on reset_index():
The reset_index() method in Pandas is used to reset the index of a DataFrame or a Series. In this case, you're calling the reset_index() method on the DataFrame df.

The first parameter drop=True indicates that the original index of the DataFrame should be dropped and not added as a new column in the resulting DataFrame.

The second parameter inplace=True indicates that the original DataFrame should be modified in place, rather than creating a copy of the DataFrame with the new index.

Therefore, the resulting output of this code will be a modified version of the df DataFrame where the original index has been reset and the modified DataFrame is saved in memory.

#### Correct the poster path :

In [247]:
base_poster_url = 'http://image.tmdb.org/t/p/w185/'
df.poster_path = "<img src='" + base_poster_url + df.poster_path + "' style='height:100px;'>"

In [248]:
df['poster_path'][0]

"<img src='http://image.tmdb.org/t/p/w185//rhIRbceoE9lR4veEXuwCC2wARtG.jpg' style='height:100px;'>"

## Save the dataframe to new file

In [ ]:
df.to_csv(movie_